klib/syslog.c

```diff
@@ -0,0 +1,338 @@
+    timestamp t = kfuncs.now(CLOCK_ID_REALTIME);
+    u64 seconds = sec_from_timestamp(t);
+    struct tm tm;
+    kfuncs.gmtime_r(&seconds, &tm);
+    kfuncs.rsnprintf(pb->payload, syslog.hdr_len + 1,
+        "<" __XSTRING(SYSLOG_PRIORITY) ">1 %d-%02d-%02dT%02d:%02d:%02d.%06dZ %s %b - - - ",
+        1900 + tm.tm_year, 1 + tm.tm_mon, tm.tm_mday, tm.tm_hour, tm.tm_min, tm.tm_sec,
+        usec_from_timestamp(t) - seconds * MILLION, syslog.local_ip, syslog.program);
+    kfuncs.runtime_memcpy(pb->payload + syslog.hdr_len, s, count);
+    kfuncs.udp_sendto(syslog.udp_pcb, pb, &syslog.server_ip, SYSLOG_UDP_PORT);
+    kfuncs.pbuf_free(pb);
+}
+
+int init(void *md, klib_get_sym get_sym, klib_add_sym add_sym)
+{
+    void (*rprintf)(const char *format, ...);
+    if (!(rprintf = get_sym("rprintf")))
+        return KLIB_INIT_FAILED;
+    tuple (*get_root_tuple)(void) = get_sym("get_root_tuple");
+    symbol (*intern)(string name) = get_sym("intern");
+    value (*get)(value e, symbol a) = get_sym("get");
+    int (*runtime_memcmp)(const void *a, const void *b, bytes len) = get_sym("runtime_memcmp");
+    u64 (*fsfile_get_length)(fsfile f) = get_sym("fsfile_get_length");
+    void *(*get_kernel_heaps)(void) = get_sym("get_kernel_heaps");
+    struct udp_pcb *(*udp_new)(void) = get_sym("udp_new");
+    boolean (*attach_console_driver)(struct console_driver *driver) =
+            get_sym("attach_console_driver");
+    if (!get_root_tuple || !intern || !get || !runtime_memcmp || !fsfile_get_length ||
+            !get_kernel_heaps || !udp_new || !attach_console_driver ||
+            !(kfuncs.fsfile_open_or_create = get_sym("fsfile_open_or_create")) ||
+            !(kfuncs.fsfile_get_writer = get_sym("fsfile_get_writer")) ||
+            !(kfuncs.allocate_sg_list = get_sym("allocate_sg_list")) ||
+            !(kfuncs.sg_list_tail_add = get_sym("sg_list_tail_add")) ||
+            !(kfuncs.deallocate_sg_list = get_sym("deallocate_sg_list")) ||
+            !(kfuncs.runtime_memcpy = get_sym("runtime_memcpy")) ||
+            !(kfuncs.register_timer = get_sym("kern_register_timer")) ||
+            !(kfuncs.timm_dealloc = get_sym("timm_dealloc")) ||
+            !(kfuncs.fs_rename = get_sym("fs_rename")) ||
+            !(kfuncs.dns_gethostbyname = get_sym("dns_gethostbyname")) ||
+            !(kfuncs.netif_get_default = get_sym("netif_get_default")) ||
+            !(kfuncs.ipaddr_ntoa_r = get_sym("ipaddr_ntoa_r")) ||
+            !(kfuncs.pbuf_alloc = get_sym("pbuf_alloc")) ||
+            !(kfuncs.pbuf_free = get_sym("pbuf_free")) ||
+            !(kfuncs.now = get_sym("now")) ||
+            !(kfuncs.gmtime_r = get_sym("gmtime_r")) ||
+            !(kfuncs.rsnprintf = get_sym("rsnprintf")) ||
+            !(kfuncs.udp_sendto = get_sym("udp_sendto"))) {
+        rprintf("syslog: kernel symbols not found\n");
+        return KLIB_INIT_FAILED;
+    }
+    tuple root = get_root_tuple();
+    buffer output = get(root, sym(syslog_output));
```

Comment:
can use get_string unless you prefer the distinct error messages below

Reply:
get_string() cannot be used from klib code, because it's an inline function that calls a non-inline function.